source/fe/fe_tools_interpolate.cc

```diff
@@ -712,6 +719,1426 @@ namespace FETools
+                                       tmp);
+          // there is no new cell_data
+          // this process needs for computing this cell,
+          // store cell_data in the list of
+          // computed cells and erase this cell
+          // from the list of cells to compute
+          if (tmp.size () == 0)
+            {
+              cell_data_insert (cells_to_compute[pos], computed_cells);
+              cells_to_compute.erase (cells_to_compute.begin () + pos);
+            }
+          else
+            {
+              for (unsigned int i=0; i<tmp.size (); ++i)
+                cell_data_insert (tmp[i], new_needs);
+            }
+        }
+
+      // search recursively over this tree
+      if (p4est_has_children)
+        {
+          typename dealii::internal::p4est::types<dim>::quadrant
+          p4est_child[GeometryInfo<dim>::max_children_per_cell];
+
+          dealii::internal::p4est::init_quadrant_children<dim> (p4est_cell,
+                                                                p4est_child);
+
+          for (unsigned int c=0; c<GeometryInfo<dim>::max_children_per_cell; ++c)
+            {
+              compute_cells_in_tree_recursively (forest,
+                                                 tree,
+                                                 tree_index,
+                                                 dealii_cell->child(c),
+                                                 p4est_child[c],
+                                                 u,
+                                                 cells_to_compute,
+                                                 computed_cells,
+                                                 new_needs);
+            }
+        }
+    }
+
+
+
+    template <int dim,int spacedim>
+    void
+    ExtrapolateImplementation<dim,spacedim>::
+    send_cells (const std::vector<CellData>  &cells_to_send,
+                std::vector<CellData>        &received_cells) const
+    {
+      std::vector<std::vector<char> > sendbuffers (cells_to_send.size());
+      std::vector<std::vector<char> >::iterator buffer = sendbuffers.begin();
+      std::vector<MPI_Request> requests (cells_to_send.size());
+      std::vector<unsigned int> destinations;
+
+      // send data
+      unsigned int idx=0;
+      for (typename std::vector<CellData>::const_iterator it=cells_to_send.begin();
+           it!=cells_to_send.end();
+           ++it, ++idx)
+        {
+          destinations.push_back (it->receiver);
+
+          it->pack_data (*buffer);
+          MPI_Isend (&(*buffer)[0], buffer->size(),
+                     MPI_BYTE,
+                     it->receiver,
+                     123,
+                     communicator,
+                     &requests[idx]);
+        }
+
+      Assert(destinations.size()==cells_to_send.size(), ExcInternalError());
+
+      std::vector<unsigned int> senders
+        = Utilities::MPI::compute_point_to_point_communication_pattern(communicator, destinations);
```

Comment:
does this pattern change in each iteration? If not, it is probably better to extract this and only compute it once.